### DISCUSSION TITLE
Import local python code takes long

### DISCUSSION THREAD

**Discussion by Omicronlawful** [PyO3/pyo3#3192] [Questions] [answered]
### Bug Description

I want to import a module i have in python, which is quite big. I imported it as seen below. I am using pyo3_asyncio to execute a function from this package which takes about 30sec to run, but the time it takes to import is 2 seconds.
This line ` let test = py.import("test")?;`takes about 2 seconds to execute. I want to compile the project later with the dependecies. Is there a way to optimize the import or load it when compiling so, it wont take the time when running?

### Steps to Reproduce

 Python::with_gil(|py| {
        let sys: &PyAny = py.import("sys")?.getattr("path")?;
        let sys: &PyList = pyo3::PyTryInto::try_into(sys).unwrap();
        sys.insert(0, path)?;
        let test = py.import("test")?;

### Your operating system and version

Arch Linux

### Your Python version (`python --version`)

Python 3.10

### Your Rust version (`rustc --version`)

rustc 1.65.0 (897e37553 2022-11-02)

### Your PyO3 version

pyo3 = { version = "0.18.2", features = ["auto-initialize"] }

### How did you install python? Did you use a virtualenv?

yay -S python310
i used venv

**Comment by adamreichold** [comment]
Are you sure this is a bug rather than a discussion? Does importing the module from Rust is measurably slower than import it from Python?

**Comment by Omicronlawful** [comment]
in python i can just execute the code. I am looking for a way to load python code without the time to wait. The execution time is just slighlty worse in rust(2%)

**Reply by adamreichold** [reply, marked as answer]
I don't think this is possible in general as importing a Python module can execute arbitrary code, so there is most likely no easy fix.

As this does depend on the specific module in question, you could start with trying to find out what exactly is taking too long. Some Linux distributions try to improve load times by shipping pre-compiled byte code (`.pyc` files, c.f. https://docs.python.org/3/library/py_compile.html) but that only helps if the time is actually spent parsing source code into byte code.
